text/0000-cargo-workspace-deduplicate.md

```diff
@@ -0,0 +1,633 @@
+dependencies in `[workspace.dependencies]` are depended on by some crate, even
+if no crate actually references an entry in `[workspace.dependencies]`. This
+means that if a crate uses an entry in `[workspace.dependencies]` it's
+guaranteed to have an entry in the lock file indicating what its dependencies
+should be.
+
+Note that for entries in `[workspace.dependencies]` which aren't used by any
+crates in the workspace will likely trigger a warning, however, so users can
+continue to prune accidentally unused entries.
+
+## Effect on `cargo metadata`
+
+Executing `cargo metadata` to learn about a crate graph will implicitly perform
+all subsitution defined in this proposal. Consumers of `cargo metadata` will
+continue to get the same output they got before this proposal, meaning that
+implicit substitutions, if any, will be invisible to users of `cargo metadata`.
+
+## Effect on `cargo read-manifest`
+
+Similar to `cargo metadata`, the `cargo read-manifest` command will perform all
+necessary subsitutions when presenting the output as JSON.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal significantly complicates the process of interpreting a
+`Cargo.toml`. One of the major purposes of using TOML to specify a crate
+manifest was to make it easy for other tools to parse Cargo manifests and work
+with them. This not only includes Rust-based tools but also tools in other
+languages if necessary. Previously a TOML parser for your language was all you
+really needed, but this proposal is adding a layer of indirection on top of TOML
+where you have to interpret multiple manifests to figure out what one means. For
+example you can no longer quickly and easily be guaranteed to parse the version
+of a package, but you might have to go find the workspace root or other crates
+to figure that out. Workspace discovery and membership is pretty nontrivial so
+non-Cargo based tools will have a difficult time *not* using Cargo to figure out
+a full elaborated form of a manifest 100% of the time.
+
+This proposal also extends `Cargo.toml` with changes that will break any
+existing tools which assume a particular format of `Cargo.toml`. For example if
+a tool expects `package.vesion` to be a `String` that runs a risk of being
+broken in the future due to the ability to specify a table there instead.
+
+Additionally this proposal complicates a reader's understanding of `Cargo.toml`.
+While verbose for maintainers having duplication of information is actually
+quite nice for readers of `Cargo.toml` because you don't have to chase anything
+else down to figure out what a dependency is. If this proposal is implemented
+then whenever you see `foo = { workspace = true }` you've got to go consult
+something else to figure out what the dependency actually is. This layer of
+indirection can cause surprise for readers or otherwise add a speed-bump to
+understanding the contents of a manifest.
```

Comment:
Would it be possible to have a way to view the indirection that occurred in some way?

Reply:
Hm could you elaborate a bit? Do you mean for example that `cargo metadata` would indicate the indirection happened? Additionally what are you thinking this'd be used for?

Reply:
Mostly trying to think of ways to help explain that surprise. If a user says "hey, where did my <homepage key> come from", is there some way we can provide an answer?

Perhaps `cargo metadata --explain` or something (not wanting to get too deep into implementation bikeshedding) could explain it.

Reply:
Ah ok I think I understand, thanks! I'd personally be somewhat hesitatnt to add that sort of infrastructure unless there's motivation from users. I largely just manufactured this drawback since it seems like a possible drawback, but in reality I feel like it'll be a quite minor drawback that won't come into play all that often (and so wouldn't want to accidentally over-rotate the implementation trying to fix it in response)